File: Dev/TemplateFactory/tools/render_runtime_plugin_profile.js
#!/usr/bin/env node

const fs = require("fs");
const path = require("path");
const { execFileSync } = require("child_process");

const ROOT = process.cwd();
const SOURCE = "Dev/TemplateFactory/modules/runtime_plugin_profile.yaml";
const errors = [];

function repoPath(relPath) {
    return path.join(ROOT, relPath);
}

function loadYaml(relPath) {
    try {
        const script = [
            "import json, sys, yaml",
            "with open(sys.argv[1], encoding='utf-8') as handle:",
            "    data = yaml.safe_load(handle) or {}",
            "print(json.dumps(data, ensure_ascii=False))"
        ].join("\n");
        const stdout = execFileSync("python3", ["-c", script, repoPath(relPath)], {
            encoding: "utf8",
            maxBuffer: 8 * 1024 * 1024
        });
        return JSON.parse(stdout);
    } catch (error) {
        errors.push(`${relPath}: YAML non leggibile (${error.message})`);
        return {};
    }
}

function asArray(value) {
    return Array.isArray(value) ? value : [];
}

function normalizeList(value) {
    return asArray(value).map(item => String(item).trim()).filter(Boolean);
}

function requireId(source, expected, relPath) {
    if (source.id !== expected) errors.push(`${relPath}: id atteso ${expected}`);
}

function jsonText(data) {
    return `${JSON.stringify(data, null, 2)}\n`;
}

function addPlugin(target, label, data) {
    const normalizedLabel = String(label ?? "").trim();
    const id = String(data?.id ?? "").trim();
    if (!normalizedLabel || !id) return;
    target[normalizedLabel] = {
        id,
        label: normalizedLabel,
        source: data.source ?? "community",
        function: String(data.function ?? "").trim(),
        symptom: String(data.symptom ?? data.visible_failure ?? data.function ?? "").trim(),
        manual_action: String(data.manual_action ?? "").trim()
    };
}

const profileSource = loadYaml(SOURCE);
requireId(profileSource, "runtime_plugin_profile", SOURCE);

const sourcePaths = profileSource.sources ?? {};
const pluginMatrixPath = String(sourcePaths.plugin_matrix ?? "Dev/TemplateFactory/modules/plugin_matrix.yaml");
const pluginContractsPath = String(sourcePaths.plugin_contracts ?? "Dev/TemplateFactory/modules/plugin_contracts.yaml");
const workflowsPath = String(sourcePaths.workflows ?? "Dev/TemplateFactory/modules/workflows.yaml");
const outputPath = String(profileSource.output ?? "z.automazioni/data/runtime/plugin_profile.json");

const pluginMatrix = loadYaml(pluginMatrixPath);
const pluginContracts = loadYaml(pluginContractsPath);
const workflows = loadYaml(workflowsPath);
requireId(pluginMatrix, "plugin_matrix", pluginMatrixPath);
requireId(pluginContracts, "plugin_contracts", pluginContractsPath);
requireId(workflows, "workflows", workflowsPath);

const matrixById = new Map(asArray(pluginMatrix.plugins).map(plugin => [String(plugin.id ?? "").trim(), plugin]));
const contractById = new Map(asArray(pluginContracts.plugins).map(plugin => [String(plugin.id ?? "").trim(), plugin]));
const manualActions = profileSource.manual_actions ?? {};
const defaultManualAction = String(profileSource.defaults?.manual_action ?? "").trim();
const pluginsByLabel = {};

for (const [pluginId, matrixPlugin] of matrixById.entries()) {
    if (!pluginId) continue;
    const contractPlugin = contractById.get(pluginId) ?? {};
    const base = {
        id: pluginId,
        source: "community",
        function: matrixPlugin.function ?? contractPlugin.local_scope ?? "",
        symptom: contractPlugin.visible_failure ?? matrixPlugin.function ?? "",
        manual_action: manualActions[matrixPlugin.name] ?? manualActions[contractPlugin.name] ?? manualActions[pluginId] ?? defaultManualAction
    };
    addPlugin(pluginsByLabel, matrixPlugin.name, base);
    if (contractPlugin.name && contractPlugin.name !== matrixPlugin.name) {
        addPlugin(pluginsByLabel, contractPlugin.name, {
            ...base,
            manual_action: manualActions[contractPlugin.name] ?? base.manual_action
        });
    }
}

for (const corePlugin of asArray(profileSource.core_plugins)) {
    const label = String(corePlugin.label ?? "").trim();
    addPlugin(pluginsByLabel, label, {
        id: corePlugin.id,
        source: corePlugin.source ?? "core",
        function: corePlugin.function,
        symptom: corePlugin.symptom,
        manual_action: manualActions[label] ?? corePlugin.manual_action ?? defaultManualAction
    });
}

for (const [label, action] of Object.entries(manualActions)) {
    if (!pluginsByLabel[label]) {
        errors.push(`${SOURCE}: manual_actions.${label} non corrisponde a nessun plugin runtime`);
        continue;
    }
    pluginsByLabel[label].manual_action = String(action).trim();
}

const workflowPlugins = {};
for (const [workflowId, workflow] of Object.entries(workflows.workflows ?? {})) {
    const requiredPlugins = normalizeList(workflow.required_plugins);
    if (!requiredPlugins.length) continue;
    workflowPlugins[workflowId] = requiredPlugins;
    for (const label of requiredPlugins) {
        if (!pluginsByLabel[label]) {
            errors.push(`${workflowsPath}: ${workflowId}.required_plugins contiene plugin non profilato (${label})`);
        }
    }
}

const defaultWorkflowPlugins = normalizeList(profileSource.defaults?.workflow_plugins);
for (const label of defaultWorkflowPlugins) {
    if (!pluginsByLabel[label]) {
        errors.push(`${SOURCE}: defaults.workflow_plugins contiene plugin non profilato (${label})`);
    }
}

const profile = {
    generated_from: {
        runtime_plugin_profile: SOURCE,
        plugin_matrix: pluginMatrixPath,
        plugin_contracts: pluginContractsPath,
        workflows: workflowsPath
    },
    default_workflow_plugins: defaultWorkflowPlugins,
    default_manual_action: defaultManualAction,
    plugins_by_label: Object.fromEntries(Object.entries(pluginsByLabel).sort(([left], [right]) => left.localeCompare(right))),
    workflow_plugins: Object.fromEntries(Object.entries(workflowPlugins).sort(([left], [right]) => left.localeCompare(right)))
};

if (!errors.length) {
    const target = repoPath(outputPath);
    const expected = jsonText(profile);
    if (process.argv.includes("--check")) {
        const current = fs.existsSync(target) ? fs.readFileSync(target, "utf8") : "";
        if (current !== expected) {
            errors.push(`${outputPath} non allineato a ${SOURCE}`);
        }
    } else {
        fs.mkdirSync(path.dirname(target), { recursive: true });
        fs.writeFileSync(target, expected, "utf8");
    }
}

if (errors.length) {
    console.error("Runtime plugin profile non valido:");
    for (const error of errors) console.error(`- ${error}`);
    process.exit(1);
}

console.log(`Runtime plugin profile OK: ${Object.keys(profile.plugins_by_label).length} label plugin, ${Object.keys(profile.workflow_plugins).length} workflow.`);

File: Dev/Tools/python/render_runtime_plugin_profile.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path
from typing import Any

import yaml

sys.dont_write_bytecode = True

ROOT = Path(__file__).resolve().parents[3]
SOURCE = "Dev/Source/YAML/json/runtime_plugin_profile.yaml"


def load_yaml(rel_path: str, errors: list[str]) -> dict[str, Any]:
    path = ROOT / rel_path
    try:
        data = yaml.safe_load(path.read_text(encoding="utf-8")) or {}
    except Exception as exc:  # noqa: BLE001
        errors.append(f"{rel_path}: YAML non leggibile ({exc})")
        return {}
    if not isinstance(data, dict):
        errors.append(f"{rel_path}: root YAML non valida")
        return {}
    return data


def require_id(source: dict[str, Any], expected: str, rel_path: str, errors: list[str]) -> None:
    if source.get("id") != expected:
        errors.append(f"{rel_path}: id atteso {expected}")


def as_list(value: Any) -> list[Any]:
    return value if isinstance(value, list) else []


def normalize_list(value: Any) -> list[str]:
    return [text for text in (str(item).strip() for item in as_list(value)) if text]


def add_plugin(target: dict[str, dict[str, str]], label: Any, data: dict[str, Any]) -> None:
    normalized_label = str(label or "").strip()
    plugin_id = str(data.get("id") or "").strip()
    if not normalized_label or not plugin_id:
        return
    target[normalized_label] = {
        "id": plugin_id,
        "label": normalized_label,
        "source": data.get("source") or "community",
        "function": str(data.get("function") or "").strip(),
        "symptom": str(data.get("symptom") or data.get("visible_failure") or data.get("function") or "").strip(),
        "manual_action": str(data.get("manual_action") or "").strip(),
    }


def json_text(data: dict[str, Any]) -> str:
    return json.dumps(data, ensure_ascii=False, indent=2) + "\n"


def js_locale_key(value: str) -> tuple[str, tuple[int, ...]]:
    # Mirrors the simple labels used by Node's default localeCompare: compare
    # case-insensitively first, then keep lowercase before uppercase on ties.
    return (value.casefold(), tuple(0 if char.islower() else 1 if char.isupper() else 2 for char in value))


def build_profile(errors: list[str]) -> tuple[dict[str, Any], str]:
    profile_source = load_yaml(SOURCE, errors)
    require_id(profile_source, "runtime_plugin_profile", SOURCE, errors)

    source_paths = profile_source.get("sources") if isinstance(profile_source.get("sources"), dict) else {}
    plugin_matrix_path = str(source_paths.get("plugin_matrix") or "Dev/Source/YAML/json/plugin_matrix.yaml")
    plugin_contracts_path = str(source_paths.get("plugin_contracts") or "Dev/Source/YAML/canonical/plugin_contracts.yaml")
    workflows_path = str(source_paths.get("workflows") or "Dev/Source/YAML/json/workflows.yaml")
    output_path = str(profile_source.get("output") or "z.automazioni/data/runtime/plugin_profile.json")

    plugin_matrix = load_yaml(plugin_matrix_path, errors)
    plugin_contracts = load_yaml(plugin_contracts_path, errors)
    workflows = load_yaml(workflows_path, errors)
    require_id(plugin_matrix, "plugin_matrix", plugin_matrix_path, errors)
    require_id(plugin_contracts, "plugin_contracts", plugin_contracts_path, errors)
    require_id(workflows, "workflows", workflows_path, errors)

    matrix_by_id = {
        str(plugin.get("id") or "").strip(): plugin
        for plugin in as_list(plugin_matrix.get("plugins"))
        if isinstance(plugin, dict)
    }
    contract_by_id = {
        str(plugin.get("id") or "").strip(): plugin
        for plugin in as_list(plugin_contracts.get("plugins"))
        if isinstance(plugin, dict)
    }
    manual_actions = profile_source.get("manual_actions") if isinstance(profile_source.get("manual_actions"), dict) else {}
    defaults = profile_source.get("defaults") if isinstance(profile_source.get("defaults"), dict) else {}
    default_manual_action = str(defaults.get("manual_action") or "").strip()
    plugins_by_label: dict[str, dict[str, str]] = {}

    for plugin_id, matrix_plugin in matrix_by_id.items():
        if not plugin_id:
            continue
        contract_plugin = contract_by_id.get(plugin_id) or {}
        base = {
            "id": plugin_id,
            "source": "community",
            "function": matrix_plugin.get("function") or contract_plugin.get("local_scope") or "",
            "symptom": contract_plugin.get("visible_failure") or matrix_plugin.get("function") or "",
            "manual_action": (
                manual_actions.get(matrix_plugin.get("name"))
                or manual_actions.get(contract_plugin.get("name"))
                or manual_actions.get(plugin_id)
                or default_manual_action
            ),
        }
        add_plugin(plugins_by_label, matrix_plugin.get("name"), base)
        if contract_plugin.get("name") and contract_plugin.get("name") != matrix_plugin.get("name"):
            alternate = {**base, "manual_action": manual_actions.get(contract_plugin.get("name")) or base["manual_action"]}
            add_plugin(plugins_by_label, contract_plugin.get("name"), alternate)

    for core_plugin in as_list(profile_source.get("core_plugins")):
        if not isinstance(core_plugin, dict):
            continue
        label = str(core_plugin.get("label") or "").strip()
        add_plugin(
            plugins_by_label,
            label,
            {
                "id": core_plugin.get("id"),
                "source": core_plugin.get("source") or "core",
                "function": core_plugin.get("function"),
                "symptom": core_plugin.get("symptom"),
                "manual_action": manual_actions.get(label) or core_plugin.get("manual_action") or default_manual_action,
            },
        )

    for label, action in manual_actions.items():
        if label not in plugins_by_label:
            errors.append(f"{SOURCE}: manual_actions.{label} non corrisponde a nessun plugin runtime")
            continue
        plugins_by_label[label]["manual_action"] = str(action).strip()

    workflow_plugins: dict[str, list[str]] = {}
    workflow_items = workflows.get("workflows") if isinstance(workflows.get("workflows"), dict) else {}
    for workflow_id, workflow in workflow_items.items():
        workflow_data = workflow if isinstance(workflow, dict) else {}
        required_plugins = normalize_list(workflow_data.get("required_plugins"))
        if not required_plugins:
            continue
        workflow_plugins[str(workflow_id)] = required_plugins
        for label in required_plugins:
            if label not in plugins_by_label:
                errors.append(f"{workflows_path}: {workflow_id}.required_plugins contiene plugin non profilato ({label})")

    default_workflow_plugins = normalize_list(defaults.get("workflow_plugins"))
    for label in default_workflow_plugins:
        if label not in plugins_by_label:
            errors.append(f"{SOURCE}: defaults.workflow_plugins contiene plugin non profilato ({label})")

    profile = {
        "generated_from": {
            "runtime_plugin_profile": SOURCE,
            "plugin_matrix": plugin_matrix_path,
            "plugin_contracts": plugin_contracts_path,
            "workflows": workflows_path,
        },
        "default_workflow_plugins": default_workflow_plugins,
        "default_manual_action": default_manual_action,
        "plugins_by_label": {key: plugins_by_label[key] for key in sorted(plugins_by_label, key=js_locale_key)},
        "workflow_plugins": {key: workflow_plugins[key] for key in sorted(workflow_plugins, key=js_locale_key)},
    }
    return profile, output_path


def main() -> int:
    parser = argparse.ArgumentParser(description="Genera il profilo runtime dei plugin da YAML.")
    parser.add_argument("--check", action="store_true")
    args = parser.parse_args()

    errors: list[str] = []
    profile, output_path = build_profile(errors)
    if not errors:
        target = ROOT / output_path
        expected = json_text(profile)
        if args.check:
            current = target.read_text(encoding="utf-8") if target.exists() else ""
            if current != expected:
                errors.append(f"{output_path} non allineato a {SOURCE}")
        else:
            target.parent.mkdir(parents=True, exist_ok=True)
            target.write_text(expected, encoding="utf-8")

    if errors:
        print("Runtime plugin profile non valido:", file=sys.stderr)
        for error in errors:
            print(f"- {error}", file=sys.stderr)
        return 1

    print(
        "Runtime plugin profile OK: "
        f"{len(profile.get('plugins_by_label') or {})} label plugin, "
        f"{len(profile.get('workflow_plugins') or {})} workflow."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
